rust/otap-dataflow/crates/otap/src/otap_grpc/proxy.rs

```diff
@@ -0,0 +1,939 @@
+    #[must_use]
+    pub fn has_proxy(&self) -> bool {
+        self.http_proxy.is_some() || self.https_proxy.is_some() || self.all_proxy.is_some()
+    }
+}
+
+/// Parses a proxy URL and returns (host, port, auth_header_value).
+fn parse_proxy_url(proxy_url: &str) -> Result<(String, u16, Option<String>), ProxyError> {
+    let uri: Uri = proxy_url
+        .parse()
+        .map_err(|_| ProxyError::InvalidProxyUrl(proxy_url.to_string()))?;
+
+    // Reject https:// proxy URLs - we don't support TLS to the proxy server itself
+    if uri.scheme_str() == Some("https") {
+        return Err(ProxyError::InvalidProxyUrl(format!(
+            "https:// proxy URLs are not supported (proxy URL: {}). \
+Use http:// instead - the CONNECT tunnel will still encrypt \
+traffic to the final destination for https:// targets.",
+            proxy_url
+        )));
+    }
+
+    let host = uri
+        .host()
+        .ok_or_else(|| ProxyError::InvalidProxyUrl(format!("missing host in {proxy_url}")))?
+        .to_string();
+
+    let port = uri.port_u16().unwrap_or(3128); // Default proxy port
+
+    // Extract credentials if present
+    let auth_header = if let Some(authority) = uri.authority() {
+        let auth_str = authority.as_str();
+        if let Some((user_pass, _)) = auth_str.rsplit_once('@') {
+            let encoded = BASE64_STANDARD.encode(user_pass);
+            Some(format!("Basic {}", encoded))
+        } else {
+            None
+        }
+    } else {
+        None
+    };
+
+    Ok((host, port, auth_header))
+}
+
+/// Establishes an HTTP CONNECT tunnel through a proxy.
+///
+/// This function connects to the proxy, sends an HTTP CONNECT request,
+/// and returns the TCP stream once the tunnel is established.
+#[cfg(test)]
+async fn http_connect_tunnel(
+    proxy_host: &str,
+    proxy_port: u16,
+    target_host: &str,
+    target_port: u16,
+) -> Result<TcpStream, ProxyError> {
+    // Connect to the proxy
+    let stream = TcpStream::connect((proxy_host, proxy_port)).await?;
+
+    http_connect_tunnel_on_stream(stream, target_host, target_port, None).await
+}
+
+async fn http_connect_tunnel_on_stream(
+    stream: TcpStream,
+    target_host: &str,
+    target_port: u16,
+    proxy_auth: Option<&str>,
+) -> Result<TcpStream, ProxyError> {
+    // Send HTTP CONNECT request
+    // Note: We use "Connection: Keep-Alive" instead of "Proxy-Connection" as the latter
+    // is non-standard, although widely supported. Modern proxies should respect "Connection".
+
+    // Handle IPv6 addresses in Host header (must be bracketed)
+    let formatted_target = if target_host.parse::<std::net::Ipv6Addr>().is_ok() {
+        format!("[{}]", target_host)
+    } else {
+        target_host.to_string()
+    };
+
+    let mut connect_request = format!(
+        "CONNECT {formatted_target}:{target_port} HTTP/1.1\r\n\
+         Host: {formatted_target}:{target_port}\r\n\
+         Connection: Keep-Alive\r\n"
+    );
```

Comment:
Adding a User-Agent header will make this code more universal. Some proxies may block requests that don't have a user agent defined.